### PR TITLE
Implement subscriber persistence, Unit tests

### PR DESCRIPTION
**Pull request by tabman83**
- extend the existing repository codebase
- base the storage on subscribers rather than endpoints
- unit tests

### CODE REVIEW

File: src/CaptainHook.Api/Controllers/EventsController.cs [hunk 1]
```diff
@@ -35,6 +35,8 @@ public EventsController(IMediator mediator, IBigBrother bigBrother)
         public async Task<IActionResult> PutWebhook([FromRoute] string eventName, [FromRoute] string subscriberName, [FromBody] EndpointDto dto)
         {
             var request = new UpsertWebhookRequest(eventName, subscriberName, dto);
+
+            await Task.CompletedTask;
```
Comment (mszymeshop) [inline src/CaptainHook.Api/Controllers/EventsController.cs:39]: temporary?

File: src/CaptainHook.Domain/ValueObjects/SubscriberId.cs [hunk 2]
```diff
@@ -24,5 +24,10 @@ public override string ToString()
         {
             return _id;
         }
+
+        public static implicit operator string(SubscriberId subscriberId)
+        {
+            return subscriberId.ToString();
```
Comment (mszymeshop) [inline src/CaptainHook.Domain/ValueObjects/SubscriberId.cs:30]: handle nulls

File: src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs [hunk 3]
```diff
@@ -77,5 +47,17 @@ internal class EndpointDocument
         /// </summary>
         [JsonProperty("eventName")]
         public string EventName { get; set; }
+
+        /// <summary>
+        /// Subscriber endpoints
+        /// </summary>
+        [JsonProperty(PropertyName = "endpoints")]
+        public EndpointSubdocument[] Endpoints { get; set; }
```
Comment (mszymeshop) [inline src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs:55]: Should we talk about distinguishment between dlq, webhook, callback before the implementation is needed?

File: src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs [hunk 4]
```diff
@@ -77,5 +47,17 @@ internal class EndpointDocument
...
+
+        /// <summary>
+        /// Webhook selector
+        /// </summary>
+        [JsonProperty("webhookSelectionRule")]
+        public string WebhookSelectionRule { get; set; }
```
Comment (mszymeshop) [inline src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs:61]: What is the example usage of Rule?

File: src/Tests/CaptainHook.Application.Tests/ValidatorPipelineBehaviorTests.cs [hunk 5]
```diff
@@ -63,6 +63,7 @@ private class DivideRequestHandler : IRequestHandler<DivideRequest, OperationRes
         {
             public async Task<OperationResult<DivideResult>> Handle(DivideRequest request, CancellationToken cancellationToken)
             {
+                await Task.CompletedTask;
```
Comment (mszymeshop) [inline src/Tests/CaptainHook.Application.Tests/ValidatorPipelineBehaviorTests.cs:66]: ?

File: src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs [hunk 6]
```diff
@@ -1,28 +1,42 @@
...
+        public CosmosQuery BuildSelectAllSubscribers()
         {
             var query = new QueryDefinition("select * from c where c.type = @documentType")
-                .WithParameter("@documentType", EndpointDocument.Type);
+                .WithParameter("@documentType", SubscriberDocument.Type);
 
             return new CosmosQuery(query);
         }
+
+        public CosmosQuery BuildSelectSubscriber(SubscriberId subscriberId)
```
Comment (mszymeshop) [inline src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs:29]: Looks like this query can be updated to be more efficient.

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 1 above)
Comment (tabman83) [reply src/CaptainHook.Api/Controllers/EventsController.cs:39]: I think so. Was not part of my code, I only added this to fix the warning in sonarcloud

File: src/Tests/CaptainHook.Application.Tests/ValidatorPipelineBehaviorTests.cs (hunk 5 above)
Comment (tabman83) [reply src/Tests/CaptainHook.Application.Tests/ValidatorPipelineBehaviorTests.cs:66]: Only added to fix the warning in SonarCloud.

File: src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs (hunk 6 above)
Comment (tabman83) [reply src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs:29]: Fixed with further commit.

File: src/CaptainHook.Domain/ValueObjects/SubscriberId.cs (hunk 2 above)
Comment (tabman83) [reply src/CaptainHook.Domain/ValueObjects/SubscriberId.cs]: Fixed with further commit.

File: src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs (hunk 3 above)
Comment (tabman83) [reply src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs:55]: The WebhookType deals with that

File: src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs (hunk 4 above)
Comment (tabman83) [reply src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs:61]: We don't have it yet. This was carried over from the old model. At the time of definition it was agreed with Darek to leave it as a string for now until we have a full definition.

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 1 above)
Comment (mszymeshop) [reply src/CaptainHook.Api/Controllers/EventsController.cs:39]: I'd much rather see Task.FromResult

File: src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs (hunk 3 above)
Comment (mszymeshop) [reply src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs:55]: ok, got it

File: src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs (hunk 6 above)
Comment (mszymeshop) [reply src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs:29]: subscriber id used as partition key?

File: src/CaptainHook.Api/Controllers/EventsController.cs (hunk 1 above)
Comment (dawidkeshop) [reply src/CaptainHook.Api/Controllers/EventsController.cs:39]: It will be reworked in the next task anyway.

Comment (tabman83) [reply src/CaptainHook.Api/Controllers/EventsController.cs:39]: You're right but this is a temporary endpoint from what I can see. Nothing is being done or returned at the moment

File: src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs (hunk 6 above)
Comment (tabman83) [reply src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs:29]: Give the way the subscriberId is build, it is also the pk.

File: src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs (hunk 4 above)
Comment (chomick) [reply src/CaptainHook.Storage.Cosmos/Models/SubscriberDocument.cs:61]: @dawidkeshop @mszymeshop @tabman83 I think we should not have it unless we need it for now - in the past that was just to highlight how the 'schema' may look but at this stage I would put only data that we are using right now

File: src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs [hunk 7]
```diff
@@ -1,28 +1,39 @@
...
+using CaptainHook.Storage.Cosmos.Models;
 using Eshopworld.Data.CosmosDb;
 using Microsoft.Azure.Cosmos;
 
 namespace CaptainHook.Storage.Cosmos.QueryBuilders
 {
     public class SubscriberQueryBuilder: ISubscriberQueryBuilder
     {
-        public CosmosQuery BuildSelectSubscribersListEndpoints(string eventName)
+        public CosmosQuery BuildSelectSubscribersList(string eventName)
```
Comment (chomick) [inline src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs:10]: @dawidkeshop @mszymeshop @tabman83 a suggestion for name change: BuildSelect(For)EventSubscribers

File: src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs [hunk 8]
```diff
@@ -1,28 +1,39 @@
...
-                .WithParameter("@documentType", EndpointDocument.Type);
+                .WithParameter("@documentType", SubscriberDocument.Type);
 
             return new CosmosQuery(query);
         }
+
+        public CosmosQuery BuildSelectSubscriber(SubscriberId subscriberId)
+        {
+            var query = new QueryDefinition(@"select * from c where c.id = @id")
+                .WithParameter("@id", subscriberId);
```
Comment (chomick) [inline src/CaptainHook.Storage.Cosmos/QueryBuilders/SubscriberQueryBuilder.cs:32]: @dawidkeshop @mszymeshop @tabman83 There seem to be duplication on how id is constructed. I think we should be using document's static GetDocumentId() method to calculate it. Also, Not certain if we need a concept of subscriber id in the domain?

File: src/CaptainHook.Domain/ValueObjects/SubscriberId.cs [hunk 9]
```diff
@@ -3,9 +3,13 @@
     public class SubscriberId
```
Comment (chomick) [inline src/CaptainHook.Domain/ValueObjects/SubscriberId.cs:3]: @dawidkeshop @mszymeshop @tabman83 I know this is different PR but do we need the concept of subscriber Id in the domain? 
If we do, is the assumption that the Id in the domain is the same as Id in the database? In such case, who owns the document identifier: repo or domain?

**Comment by tabman83** [comment]
This has been merged and we should address this comments in a further PR

**Pull request merged**